controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,684 @@
+	if reflect.DeepEqual(config.Spec.Tags, upstreamSpec.Tags) {
+		logrus.Debugf("[AKS] updateUpstreamClusterState check config are similar, exit \n")
+		return config, nil
+	}
+
+	azureAuthorizer, _ := newClientAuthorizer(config.Spec)
+	resourceGroupsClient, _ := newResourceGroupsClient(azureAuthorizer, config.Spec)
+
+	exists, _ := checkResourceGroup(ctx, resourceGroupsClient, config.Spec)
+	if !exists {
+		logrus.Infof("Resource group %s does not exist, creating", config.Spec.ResourceGroup)
+		status, err := createResourceGroup(ctx, resourceGroupsClient, config.Spec)
+		if status {
+			logrus.Infof("Resource grup [%s] updated successfully", config.Name)
+		}
+		if err != nil {
+			fmt.Errorf("Error during creating resource group %v", err)
+		}
+	}
+
+	resourceClusterClient, _ := newClustersClient(azureAuthorizer, config.Spec)
+	result, err := createOrUpdateCluster(ctx, resourceClusterClient, config.Spec)
+	if err != nil {
+		fmt.Errorf("Failed to update cluster: %v", err)
+	}
+
+	clusterState := *result.ManagedClusterProperties.ProvisioningState
+	if clusterState == "Succeeded" {
+		logrus.Infof("Cluster [%s] was updated successfully", config.Name)
+	}
+
+	logrus.Infof("Cluster config [%s] was checked", config.Spec.DisplayName)
+	return config, nil
+}
+
+func newClientAuthorizer(spec aksv1.AKSClusterConfigSpec) (autorest.Authorizer, error) {
+	authBaseURL := spec.AuthBaseURL
+	if authBaseURL == "" {
+		authBaseURL = azure.PublicCloud.ActiveDirectoryEndpoint
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(authBaseURL, spec.TenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	spToken, err := adal.NewServicePrincipalToken(*oauthConfig, spec.ClientID, spec.ClientSecret, baseURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return autorest.NewBearerAuthorizer(spToken), nil
+}
+
+func newResourceGroupsClient(authorizer autorest.Authorizer, spec aksv1.AKSClusterConfigSpec) (*resources.GroupsClient, error) {
+	if authorizer == nil {
+		newAuthorizer, err := newClientAuthorizer(spec)
+		if err != nil {
+			return nil, err
+		}
+
+		authorizer = newAuthorizer
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	client := resources.NewGroupsClientWithBaseURI(baseURL, spec.SubscriptionID)
+	client.Authorizer = authorizer
+
+	return &client, nil
+}
+
+func createResourceGroup(ctx context.Context, groupsClient *resources.GroupsClient, spec aksv1.AKSClusterConfigSpec) (bool, error) {
+	future, err := groupsClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		resources.Group{
+			Name:	  to.StringPtr(spec.ResourceGroup),
+			Location: to.StringPtr(spec.ResourceLocation),
+		})
+
+	if err != nil {
+		return false, fmt.Errorf("[AKS] error creating resource group '%s': %v", spec.ResourceGroup, err)
+	}
+
+	return future.StatusCode == 201, nil
```

Comment:
Why not just returned an error if this a case so we can get rid of the error return type since we will be returning an error if this is false?